packages/asset/contracts/Asset.sol

```diff
@@ -377,72 +393,112 @@ contract Asset is
+    /// @dev The next 16 bits are the asset nonce
+    /// @dev The next 40 bits are the abilities and enhancements hash
+    /// @param creator The address of the creator of the asset
+    /// @param tier The tier of the asset determined by the catalyst used to create it
+    /// @param assetNonce The nonce of the asset creator
+    /// @param revealed Whether the asset is revealed or not
+    /// @param revealHash The ipfs hash of the abilities and enhancements of the asset
+    /// @return tokenId The generated token id
     function generateTokenId(
         address creator,
         uint8 tier,
         uint16 assetNonce,
         bool revealed,
-        uint40 abilitiesAndEnhancementsHash
-    ) public view returns (uint256) {
-        /// the token id will be a uint256 with the following structure:
-        /// 0-159 bits: creator address
-        /// 160-167 bits: chain id
-        /// 168-175 bits: tier
-        /// 176-176 bits: revealed 0 | 1
-        /// 177-193 bits: creator nonce
-        /// 194-234 bits: hash of the abilities and enhancements
-        /// 235-255 bits: reserved for future use
-
-        // convert the address to uint160
+        uint40 revealHash
+    ) public view returns (uint256 tokenId) {
         uint160 creatorAddress = uint160(creator);
-        // convert the mint as revealed to uint8
         uint8 revealedUint8 = revealed ? 1 : 0;
 
-        // create the token id
-        uint256 tokenId = uint256(
-            creatorAddress |
-                (chainIndex << 160) |
-                (uint256(tier) << 168) |
-                (uint256(revealedUint8) << 176) |
-                (uint256(assetNonce) << 177) |
-                (uint256(abilitiesAndEnhancementsHash) << 194)
-        );
+        tokenId = tokenId =
+            uint256(creatorAddress) |
+            (uint256(chainIndex) << CHAIN_SHIFT) |
+            (uint256(tier) << TIER_SHIFT) |
+            (uint256(revealedUint8) << REVEALED_SHIFT) |
+            (uint256(assetNonce) << NONCE_SHIFT) |
+            (uint256(revealHash) << HASH_SHIFT);
 
         return tokenId;
     }
 
+    /// @notice Extracts the creator address from a given token id
+    /// @param tokenId The token id to extract the creator address from
+    /// @return creator The asset creator address
     function extractCreatorFromId(
         uint256 tokenId
     ) public pure returns (address creator) {
         creator = address(uint160(tokenId));
+        return creator;
     }
 
-    function extractTierFromId(uint256 tokenId) public pure returns (uint256) {
-        uint256 tier = (tokenId >> 168) & 0xFF;
+    /// @notice Extracts the chain index from a given token id
+    /// @param tokenId The token id to extract the chain index from
+    /// @return chainIdx The chain index
+    function extractChainIndexFromId(
+        uint256 tokenId
+    ) public view returns (uint8 chainIdx) {
+        chainIdx = uint8((tokenId >> CHAIN_SHIFT) & CHAIN_MASK);
```

Comment:
there's no way to do this in one operation ?

Reply:
Unfortunately not to my knowledge.